### PR TITLE
Open Google Maps if present on iOS

### PR DESCRIPTION
**Pull request by nwaughachukwuma**
On iOS, open Google Maps if present, else open Apple Maps. @brh55 this is a proposal for #25. Please see if it aligns with your ideas for the library. Thanks

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -115,13 +115,32 @@ export default function open(params) {
...
 
-export function createOpenLink({ provider, ...params }) {
+export async function createOpenLink({ provider, ...params }) {
 	// Returns a delayed async function that opens when executed
+	let mapProvider = provider;
 	if (!provider) {
-		defaultProvider = (Platform.OS === 'ios') ? 'apple' : 'google';
-	}
+		// use default provider
+		mapProvider = (Platform.OS === 'ios') ? 'apple' : 'google';
```
Comment (nwaughachukwuma) [inline index.js]: Is there any reason defaultProvider is not declared?

File: index.js [hunk 2]
```diff
@@ -115,13 +115,32 @@ export default function open(params) {
 	createOpenLink(params)();
 }
 
-export function createOpenLink({ provider, ...params }) {
+export async function createOpenLink({ provider, ...params }) {
 	// Returns a delayed async function that opens when executed
+	let mapProvider = provider;
 	if (!provider) {
-		defaultProvider = (Platform.OS === 'ios') ? 'apple' : 'google';
```
Comment (nwaughachukwuma) [inline index.js]: Is there any reason defaultProvider was used undeclared? I had to declare mapProvider at the top and use variable reassignment. I hope this is clean for you?

File: index.js [hunk 3]
```diff
@@ -115,13 +115,32 @@ export default function open(params) {
...
+					mapProvider =  'google'
+				} else { 
+					mapProvider = 'apple'
+					console.warn('Cannot open google maps. Falling back to apple maps'); 
+				}
+			} catch (error) {
+				mapProvider = 'apple'
+				console.warn('open google maps error: ', error)
+			}
+		}
```
Comment (nwaughachukwuma) [inline index.js]: Please let me know if this is clean for you

File: README.md [hunk 4]
```diff
@@ -93,7 +93,8 @@ Creates the raw link for the map.
 | *end* | `string` that geolocation can understand. | The end location that can be interpreted as a geolocation. See [Apple](https://developer.apple.com/library/archive/featuredarticles/iPhoneURLScheme_Reference/MapLinks/MapLinks.html#//apple_ref/doc/uid/TP40007899-CH5-SW1) and [Google](https://developers.google.com/maps/documentation/urls/guide#directions-action) docs for more details on how to define geolocations. | `"SOHO, New York, NY"` |
 | *navigate_mode* | `string` <br> [`preview`,`navigate`] | Determines whether map should open in preview mode or in navigate mode (with turn-by-turn navigation). <br> This parameter only works in conjunction with `end`. Platform map uses current location as `start` parameter | `"navigate"` <br> *default: `"preview"`* |
 
-**Note:** Combining query with latitude and longitude will override the query parameter.
+**Note 1:** Combining query with latitude and longitude will override the query parameter.
+**Note 2:** To use google maps on iOS, declare the URL scheme used by the API in your Info.plist file. Follow the step [here](https://developers.google.com/maps/documentation/ios-sdk/start#step_6_declare_the_url_schemes_used_by_the_api)
```
Comment (nwaughachukwuma) [inline README.md]: Would you rather the notes were numbered differently?

File: index.js [hunk 5]
```diff
@@ -112,19 +112,36 @@ export const createQueryParameters = ({
 };
 
 export default function open(params) {
-	createOpenLink(params)();
+	createOpenLink(params) //()
 }
```
Comment (nwaughachukwuma) [inline index.js]: Due to the await call to Linking.canOpenURL(...), I had to make `createOpenLink` an async function and return an await. This creates the delayed effect you had made in the first place. I hope this is OK?

File: index.js [hunk 6]
```diff
@@ -112,19 +112,39 @@ export const createQueryParameters = ({
...
+					}
+				} catch (error) {
+					mapProvider = 'apple'
+					console.warn('Open google maps error: ', error)
+				}
+			}
+		}
+		// Allow override provider, otherwise use the default provider
+		const mapLink = createMapLink({ provider: mapProvider, ...params });
+		return await Linking.openURL(mapLink).catch(err => console.error('An error occurred', err));
```
Comment (nwaughachukwuma) [inline index.js]: I had to put all the method body in the delayed function, in order to achieve the same functionality as before for `createOpenLink`. I have tested this and it works on both platforms

File: README.md (hunk 4 above)
Comment (brh55) [reply README.md]: You shouldn't have to declare a URL scheme, the Google URL is cross-platform supported, and not using a deep link / custom URL scheme.

**Comment by brh55** [comment]
I think the original PR addresses the main issue, it keeps the library JS only and simple. I would say this is adding more complexity and is duplicating lots of efforts already provided by other libraries.

**Comment by nwaughachukwuma** [comment]
I see your point @brh55, but the changes made have not altered the overall behavior and purpose of the library. The logic added, as shown below, only check if the preferred provider is `google` on iOS. This can be very useful for some users like #25.
```
if (Platform.OS === 'ios' && provider === 'google') {
    // Device is iOS with google maps as provider
    try {
        // check that device can open google maps. Else default to apple maps
        const canOpen = await Linking.canOpenURL('comgooglemaps://?center=40.765819,-73.975866');
        if (canOpen) { 
            mapProvider = 'google'
        } else { 
            throw new Error('Cannot open google maps, falling back to apple maps');
        }
    } catch (error) {
        mapProvider = 'apple'
        console.warn('Open google maps error: ', error)
    }
}
```
I have these changes introduced in a production app using patch-package and it's awesome. But you have the final say really. Let me know what you think

**Pull request closed without merging**